managementAgent.js

```diff
@@ -1,59 +1,159 @@
-const WebSocket = require('ws');
 const Uuid = require('uuid');
+const WebSocket = require('ws');
+const metadata = require('./lib/metadata/wrapper');
 
 const logger = require('./lib/utilities/logger');
 const { initManagement } = require('./lib/management');
 const _config = require('./lib/Config').config;
+const { managementAgentMessageType } = require('./lib/management/agentClient');
+const { addOverlayMessageListener } = require('./lib/management/push');
+
+
+// TODO: auth?
+
+const CHECK_BROKEN_CONNECTIONS_FREQUENCY_MS = 15000;
+
+
+class ManagementAgentServer {
+    constructor() {
+        this.port = _config.managementAgent.port || 8010;
+        this.wss = null;
+        this.loadedOverlay = null;
+
+        process.on('SIGINT', this.stop.bind(this));
+        process.on('SIGHUP', this.stop.bind(this));
+        process.on('SIGQUIT', this.stop.bind(this));
+        process.on('SIGTERM', this.stop.bind(this));
+        process.on('SIGPIPE', () => {});
+    }
 
-process.env.REPORT_TOKEN = process.env.REPORT_TOKEN
-                           || _config.reportToken
-                           || Uuid.v4();
+    start(_cb) {
+        const cb = _cb || function noop() {};
 
-setTimeout(() => {
-    initManagement(logger.newRequestLogger());
-}, 5000);
+        /* Define REPORT_TOKEN env variable needed by the management
+         * module. */
+        process.env.REPORT_TOKEN = process.env.REPORT_TOKEN
+          || _config.reportToken
+          || Uuid.v4();
 
-setTimeout(() => {
-    managementAgentWS();
-}, 6);
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                return cb(new Error(msg));
```

Comment:
Real `Error` objects usually don't print well in our json logs so just forwarding the error we get to the callback is safer. A log before that saying `log.error('failed to setup metadata', { error })` would be enough to track

Reply:
done